### PR TITLE
EFA-618 Use new passphrase logic

### PR DESCRIPTION
**Pull request by ig-garcia**

### CODE REVIEW

File: k9mail/src/main/java/com/fsck/k9/helper/AppUpdater.kt [hunk 1]
```diff
@@ -84,25 +84,11 @@ constructor(
...
         }
-        if (needsUpdateFor(v3_1_11)) {
-            v3111Update(storage, editor)
-        }
         if (needsUpdateFor(v3_1_14)) {
             v3114Update(editor)
         }
     }
 
-    private fun v3111Update(storage: Storage, editor: StorageEditor) {
```
Comment (Stereodark) [inline k9mail/src/main/java/com/fsck/k9/helper/AppUpdater.kt:95]: why was this removed? Because now we don't handle the passphrase for new keys anymore as in the first proposal? Like No need for the flag?

File: k9mail/src/main/java/security/planck/ui/passphrase/PassphraseDialogMode.kt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 package security.planck.ui.passphrase
 
 enum class PassphraseDialogMode {
-    MANAGE, UNLOCK
+    MANAGE, UNLOCK,
```
Comment (Stereodark) [inline k9mail/src/main/java/security/planck/ui/passphrase/PassphraseDialogMode.kt:4]: do we need the extra , without any other option in the enum?

File: k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+import security.planck.ui.passphrase.PassphraseViewModel
+import security.planck.ui.passphrase.models.AccountTextFieldState
+import security.planck.ui.passphrase.models.PassphraseState
+import security.planck.ui.passphrase.models.PassphraseVerificationStatus
+import security.planck.ui.passphrase.models.TextFieldState
+import security.planck.ui.passphrase.models.TextFieldStateContract
+import javax.inject.Inject
+
+private const val NEW_PASSPHRASE_INDEX = 0
+private const val NEW_PASSPHRASE_CONFIRMATION_INDEX = 1
```
Comment (Stereodark) [inline k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt:27]: isn't there an option to work with maps instead of a positional array? I don't know the form framework, but positional logic is error prone, if a map would be available the indexes would become keys and would not need any update in the future even if you add more elements to the form

File: k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    private val controller: MessagingController,
+    private val uiCache: PlanckUIArtefactCache,
+    private val configManager: ConfigurationManager,
+    private val dispatcherProvider: DispatcherProvider,
+    formatValidator: PassphraseFormatValidator
+) : PassphraseViewModel(formatValidator) {
+    private lateinit var account: Account
+    private var manualSetup = false
+    private val newPasswordState: AccountTextFieldState
+        get() = textFieldStates[NEW_PASSPHRASE_INDEX] as AccountTextFieldState
```
Comment (Stereodark) [inline k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt:43]: the map would be possible only if this can take a key

**Review by Stereodark — COMMENTED** [review]
I did a quick check, couldn't do a thorough review due to lack of time, didn't check the test class and also not checked the sub components. The code for the logic here looks fine, form input-wise I would use a map if possible to manage input, but it's not a blocking issue. Other than that looks good to me

File: k9mail/src/main/java/com/fsck/k9/helper/AppUpdater.kt (hunk 1 above)
Comment (ig-garcia) [reply k9mail/src/main/java/com/fsck/k9/helper/AppUpdater.kt:95]: We use it but several updates on same storage key was being buggy. Also the outcome will be the same as one update overwrites the other.

File: k9mail/src/main/java/security/planck/ui/passphrase/PassphraseDialogMode.kt (hunk 2 above)
Comment (ig-garcia) [reply k9mail/src/main/java/security/planck/ui/passphrase/PassphraseDialogMode.kt:4]: Removed, thanks.

File: k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt (hunk 3 above)
Comment (ig-garcia) [reply k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt:27]: The thing is different in this case imo.
Like I knowingly added these positions on purpose because they have special behavior. So a normal passphrase input only checks for correct format, but the ones at these indexes need to also check the verificaton is matching the new passphrase.

I  believe that most probably using a list is confusing, and that the real improvement is not even using a map, but something in the direction of changing from inheritance to composition, as we have 3 screens using ViewModels inherited from same abstract class.

Comment (ig-garcia) [reply k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt:27]: It is possible that I did not fully understand your suggestion, in that case please let's talk about it.

Comment (Stereodark) [reply k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt:27]: I understand you need to use these indexes to check if the two fields have the same value. 
The point I'm making is if we have to collect a certain number of fields, in this case 2, but could be 3 or 4 or 5 in the future, you can save them into different structures.
One could be an Array, like ['stereodark', 'Mauro Labate', 'pwd123', 'pwd123']. Then you can keep for accessing the data some separate variables that index the array, such as
const val USERNAME = 0
const val FULLNAME = 1
const val PASSWORD = 2
const val VERIFYPWD = 3
etc

If you use a map you can store {'username'='Stereodark', 'fullname'='Mauro Labate', 'password'='pwd123', 'verifypwd'='pwd123'}. This way you don't have to keep additional constants (or you can if you want to keep the keys), but more importantly if someone laters adds a field, you lower the risk of him breaking the system.

If we use the first option, the array, a junior dev might be asked to add a new field to the form, date of birth, above the passwords. He doesn't notice the constants and adds it to the view and saves it, suddenly the system doesn't work anymore because index 2 is now date of birth, password is index 3 and verify password is index 4. If you had a map this could not happen as the key is inherently part of the data structure, so adding the field would force the dev to bind it to a new key. 

Anyways not a big deal in this case, as I can't think about any reasonable other field to add to the password dialog except pwd and verify pwd, but just in general I think forms are better bound to maps than arrays.

Comment (ig-garcia) [reply k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt:27]: I see. The real issue here is probably the inheritance, like there is a list in the abstract viewmodel that it implements cause we can re-use parts. Otherwise no list needed at all. But the separation of concerns is not enough now imo, should eventually be refactored to some more flexible approach like composition instead of inheritance or a mix of both.
Like I want to use some logic in an abstract class, but I ended up also having to fit to some parts (like the list) that are not optimal here. This is at least the S and the I in SOLID.

Comment (Stereodark) [reply k9mail/src/main/java/security/planck/ui/setup/CreateAccountKeysViewModel.kt:27]: yeah makes sense, that's why I was mentioning in my initial comment I would do this is possible from the upper interface, if too complex let's not go crazy.

**Review by Stereodark — APPROVED** [review]
reviewed also tests, looks good to me, the comments are resolved, I think the keys/array can be ignored for now as it would impact the upper class in the hierarchy and it would be too complex for the minor benefits
